File: parse_data.py
import re
import json

with open('raw_data.txt', 'r', encoding='utf-8') as f:
    text = f.read()

# remove standalone numbers which are page numbers
text = re.sub(r'\n\s*\d+\s*\n', '\n', text)

sections = {}

# A: Repeats
m_repeats = re.search(r'Repeats:(.*?)(?=SENTENCE BUILDS:|SENTENCES BUILDS:)', text, re.DOTALL)
if m_repeats:
    rep_text = m_repeats.group(1)
    repeats = re.findall(r'\d+\.\s*(.*?)(?=\s+\d+\.|\Z)', rep_text, re.DOTALL)
    sections['A'] = [r.replace('\n', ' ').strip() for r in repeats if len(r.strip()) > 3]

# B: Sentence Builds
m_builds = re.search(r'SENTENCE BUILDS:(.*?)(?=SENTENCE COMPLETION::)', text, re.DOTALL)
if m_builds:
    b_text = m_builds.group(1)
    builds = []
    for m in re.finditer(r'\d+\)\s*(.*?)\s*A\)\s*(.*?)(?=\s+\d+\)|SENTENCES BUILDS:|\Z)', b_text, re.DOTALL):
        prompt = m.group(1).replace('\n', ' ').strip()
        ans = m.group(2).replace('\n', ' ').strip()
        builds.append({"prompt": prompt, "answer": ans})
    sections['B'] = builds

# E: Sentence Completion
m_comp = re.search(r'SENTENCE COMPLETION::(.*?)(?=DICTATIONS)', text, re.DOTALL)
if m_comp:
    c_text = m_comp.group(1)
    # usually starts with • or number. bullet points might be \u2022 or \u2981
    c_list = re.split(r'\n\s*(?:[•\u2022\u2981]|\d+\.)\s*', '\n' + c_text)
    sections['E'] = [c.replace('\n', ' ').strip() for c in c_list if len(c.strip()) > 5]

# D: Dictations
m_dict = re.search(r'DICTATIONS(.*?)(?=PASSAGE RECONSTRUCTION)', text, re.DOTALL)
if m_dict:
    d_text = m_dict.group(1)
    d_list = re.split(r'\n\s*(?:[•\u2022\u2981]|\d+\.)\s*', '\n' + d_text)
    sections['D'] = [d.replace('\n', ' ').strip() for d in d_list if len(d.strip()) > 5]

# F: Passage Reconstruction
m_pass = re.search(r'PASSAGE RECONSTRUCTION(.*?)(?=conversation -1|RECONSTRUCTED)', text, re.DOTALL)
if m_pass:
    p_text = m_pass.group(1)
    p_list = re.split(r'\n\s*(?:[•\u2022\u2981]|\d+\)|\d+\.)\s*', '\n' + p_text)
    sections['F'] = [p.replace('\n', ' ').strip() for p in p_list if len(p.strip()) > 10]

# C: Conversations
m_conv = re.search(r'conversation -1(.*)', text, re.DOTALL | re.IGNORECASE)
if m_conv:
    conv_text = '\n1. ' + m_conv.group(1)
    c_list = re.split(r'\n\s*\d+\.\s*', conv_text)
    sections['C'] = [c.strip() for c in c_list if len(c.strip()) > 10]

with open('data/parsed_questions.json', 'w', encoding='utf-8') as f:
    json.dump(sections, f, indent=2)

print(f"Parsed A: {len(sections.get('A', []))} items")
print(f"Parsed B: {len(sections.get('B', []))} items")
print(f"Parsed C: {len(sections.get('C', []))} items")
print(f"Parsed D: {len(sections.get('D', []))} items")
print(f"Parsed E: {len(sections.get('E', []))} items")
print(f"Parsed F: {len(sections.get('F', []))} items")

File: parse_data.js
const fs = require('fs');

const textRaw = fs.readFileSync('raw_data.txt', 'utf8');
const text = textRaw.replace(/\n\s*\d+\s*\n/g, '\n');

const aStart = text.indexOf('Repeats:');
const bStart = text.search(/SENTENCE BUILDS:|SENTENCES BUILDS:/i);
const eStart = text.search(/SENTENCECOMPLETION::|SENTENCE COMPLETION::/i);
const dStart = text.indexOf('DICTATIONS');
const fStart = text.search(/PASSAGE RECONSTRUCTION|PASSAGERECONSTRUCTION/i) !== -1 ? text.search(/PASSAGE RECONSTRUCTION|PASSAGERECONSTRUCTION/i) : text.search(/RECONSTRUCTION/i);
const cStart = text.search(/conversation\s*-\s*1/i);

console.log({ aStart, bStart, eStart, dStart, fStart, cStart });

const sections = { A: [], B: [], C: [], D: [], E: [], F: [] };

if (aStart !== -1 && bStart !== -1) {
    const repText = text.substring(aStart, bStart);
    const regex = /\d+\.\s*(.*?)(?=\s+\d+\.|$)/gs;
    let m;
    while ((m = regex.exec(repText)) !== null) {
        sections.A.push(m[1].replace(/\n/g, ' ').trim());
    }
}

if (bStart !== -1 && eStart !== -1) {
    const bText = text.substring(bStart, eStart);
    const regex = /\d+\)\s*(.*?)\s*A\)\s*(.*?)(?=\s+\d+\)|SENTENCES BUILDS:|$)/gs;
    let m;
    while ((m = regex.exec(bText)) !== null) {
        sections.B.push({ prompt: m[1].replace(/\n/g, ' ').trim(), answer: m[2].replace(/\n/g, ' ').trim() });
    }
    const b2Text = bText.substring(bText.search(/SENTENCES BUILDS:/i) || 0);
    const b2Regex = /⦁\s*(.*?)(?=\s*⦁|$)/gs;
    let temp = [];
    while ((m = b2Regex.exec(b2Text)) !== null) {
        temp.push(m[1].replace(/\n/g, '').trim());
    }
    for (let i = 0; i < temp.length - 1; i += 2) {
        sections.B.push({ prompt: temp[i], answer: temp[i + 1] });
    }
}

if (eStart !== -1 && dStart !== -1) {
    const eText = text.substring(eStart, dStart);
    const regex = /(?:⦁|•|\u2022|\u2981)\s*\d*\.?\s*(.*?)(?=\s*(?:⦁|•|\u2022|\u2981)|$)/gs;
    let m;
    while ((m = regex.exec(eText)) !== null) {
        if (!m[1].includes('SENTENCECOMPLETION')) {
            const stripped = m[1].trim();
            if (stripped.length > 5) sections.E.push(stripped.replace(/\s+/g, ' '));
        }
    }
}

if (dStart !== -1 && fStart !== -1) {
    const dText = text.substring(dStart, fStart);
    const regex = /(?:⦁|•|\u2022|\u2981)\s*(.*?)(?=\s*(?:⦁|•|\u2022|\u2981)|$)/gs;
    let m;
    while ((m = regex.exec(dText)) !== null) {
        if (!m[1].includes('DICTATIONS')) {
            const stripped = m[1].trim();
            if (stripped.length > 5) sections.D.push(stripped.replace(/\s+/g, ' '));
        }
    }
}

if (fStart !== -1 && cStart !== -1) {
    const fText = text.substring(fStart, cStart);
    const cutoff = fText.search(/RECONSTRUCTED/i);
    const fPrompts = cutoff > -1 ? fText.substring(0, cutoff) : fText;
    const regex = /(?:⦁|•|\u2022|\u2981)\s*\d+\)\s*(.*?)(?=\s*(?:⦁|•|\u2022|\u2981)|$)/gs;
    let m;
    while ((m = regex.exec(fPrompts)) !== null) {
        sections.F.push(m[1].replace(/\n/g, ' ').trim());
    }
}

if (cStart !== -1) {
    let convStr = text.substring(cStart);
    const regex = /\d+\.\s*Speaker 1:\s*(.*?)(?=\s+\d+\.\s*Speaker 1:|$)/gs;
    let m;
    while ((m = regex.exec(convStr)) !== null) {
        sections.C.push('Speaker 1: ' + m[1].trim().replace(/\s+/g, ' '));
    }
}

console.log('Parsed A:', sections.A.length);
console.log('Parsed B:', sections.B.length);
console.log('Parsed C:', sections.C.length);
console.log('Parsed D:', sections.D.length);
console.log('Parsed E:', sections.E.length);
console.log('Parsed F:', sections.F.length);

fs.writeFileSync('data/question_pool.ts', `export const QUESTION_POOL = ${JSON.stringify(sections, null, 2)};`);
